flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -449,12 +460,17 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
                 }
 
                 // Checking for exceptions from previous writes
-                mailboxExecutor.submit(
+                // Notice: throwing exception in mailboxExecutor thread is not safe enough for
+                // triggering global
+                // fail over, which has been fixed in [FLINK-31305]. And using
+                // mailboxExecutor.execute() is better than
+                // mailboxExecutor.submit() since submit will swallow exceptions inside.
+                mailboxExecutor.execute(
```

Comment:
Yes, this is the actual fix. Using execute instead of submit. This bug got introduced through FLINK-31305.

Note: `submit` does not swallow exceptions. It's an anti-pattern to use `submit` without looking at the Future. The future holds the result/exception to be handled async somewhere. If we bubble up the exception the task dies without the call-site being able to properly react to it. 
So please remove the second part of your comment from "And using ...".

I will amend the javadoc of the mailbox executor to make it clear that the executor indeed does not bubble up exception in `submit`.

Reply:
I agree. My description is not exactly, if we use submit and get the future returned, the exception will also be thrown. 

I'll remove this part: "And using ...".

Reply:
@AHeise  removed now

Reply:
Good catch, thanks for fixing this!

Reply:
As a follow-up, I opened and resolved https://issues.apache.org/jira/browse/FLINK-35796.